i18n: Use `wp-components` script handle to pass locale data to `wp.i18n`

ocean90 opened the pull request:
This changes the script handle from `wp-edit-post` to `wp-components` for the `wp_add_inline_script()` call that passes the locale data to `wp.i18n`.

Fixes #5898.

**Testing instructions:**

Change language of your WordPress install and check that all components are translated.

ocean90 committed:
i18n: Use `wp-components` handle to pass locale data to `wp.i18n`.

youknowriad commented on lib/client-assets.php:
```diff
@@ -866,7 +866,7 @@ function gutenberg_editor_scripts_and_styles( $hook ) {
 	$locale_data = gutenberg_get_jed_locale_data( 'gutenberg' );
 	wp_add_inline_script(
-		'wp-edit-post',
+		'wp-components',
```
I think this doesn't solve the conceptual problem but just moves it to another level. I think the solution here would be to create a separate script for these i18n strings and add it as a dependency to all Gutenberg scripts with translatable strings. I used this technique here https://github.com/youknowriad/gcf/pull/33/files#diff-14f2377adcf109755a6ae6321d6c35afR1

ocean90 replied:
Why so complicated if you can just use wp-i18n?

youknowriad replied:
Because if you use `wp-i18n` you're adding this script to all pages where maybe it's not needed. I guess the ideal is a domain per script :) to avoid loading everything on all pages.

I can be on board with `wp-i18n` for Gutenberg (because it's Core) but not for plugins though. I also wonder how this would grow over time :).

ocean90 replied:
Mhh, you only enqueue and add inline scripts when the page needs them. Just because plugin X adds the translations for page Y, it doesn’t mean that the translations are automatically available on all pages.

youknowriad replied:
It doesn't make sense to load the `wp-components` scripts in any page without the translated strings it depends on.

It doesn't make sense to load the `wp-editor` scripts in any page without the translated strings it depends on.

If a plugin author creates a new page in WP Admin and is depending on `wp-components` because it's using one of these generic components. The only thing this plugin author should do is add a dependency on `wp-components` to its scripts where these components are used, the plugin author shouldn't have to worry about setting translated strings for this module, this is just implementation detail of the `wp-components` script.

So this means the translated strings are just a dependency for the `wp-components` like any other dependency and it should be defined at registration time and not when enqueuing the scripts.

If you decide to add this inline script at enqueue time, you'll have to do it on each page this script is loaded in.

youknowriad replied:
Again, that's the theory, I'm fine with making a compromise for Gutenberg translation strings (or Core JS translation strings in general) and just always load them in all pages. (which means adding them to `wp-i18n` at registration time)

ocean90 replied:
I'm not really following you logic here.
`wp-components`, `wp-blocks`, `wp-editor` and `wp-edit-post` are currently the only scripts that depend on `wp-i18n` . Since the strings are not separated and based on the dependency tree, `wp-components` is the common script used by all of them and therefore should be used to register the translations. Just like in this PR demonstrated.

How scripts and translations are registered with Gutenberg in core is another topic and should probably be part of the merge proposal.

youknowriad replied:
> `wp-components` is the common script used by all of them and therefore should be used to register the translations. Just like in this PR demonstrated.

That's true now but that's just a coincidence, what if we introduce a new Gutenberg module that doesn't depend on components but still requires i18n, what if we add a translated string to `wp-data` in another PR? I'd be more confident (more future proof) if we add this as to dependency of `wp-i18n`

ocean90 committed:
i18n: Use `wp-i18n` handle to pass locale data to `wp.i18n`.

youknowriad commented on lib/client-assets.php:
```diff
@@ -866,9 +866,8 @@ function gutenberg_editor_scripts_and_styles( $hook ) {
-		'wp.i18n.setLocaleData( ' . json_encode( $locale_data ) . ' );',
-		'before'
+		'wp-i18n',
+		'wp.i18n.setLocaleData( ' . json_encode( $locale_data ) . ' );'
```
Should we move this inline script to line 140 (registration time)?

ocean90 replied:
No, because `gutenberg_editor_scripts_and_styles()` gets only called if Gutenberg gets initialized while `gutenberg_register_scripts_and_styles()` is called on all pages. Also, the strings are not used by `wp-i18n`.

youknowriad replied:
That's exactly the point I was raising before, what if a plugin adds a dependency to `wp-components` or `wp-blocks` in a separate page? The strings should be loaded without any other adjustment from the plugin author.

ocean90 replied:
It's a fundamental problem with the current API. Based on https://github.com/WordPress/packages/pull/96#issuecomment-378529253, "there's no blocking here. So moving forward.". 🤷‍♂️

youknowriad replied:
@aduth Thoughts on this

aduth replied:
It will need further changes for merge. I imagine it might be the domain-per-script idea you already mentioned, ideally with some automation for assigning domain to a script bundle for string extraction and injecting the locale data inline script. Admittedly unclear how well this fits into existing localization workflows.

As for what to do here, I don't feel strongly _against_ this implementation as a short-term solution, but I agree the locale data should ideally be set as a side effect of depending on any of the internationalized Gutenberg scripts. I see this as a stubbed script upon which all of the Gutenberg bundles depend, with the inline script to set locale data. This was already mentioned, though dismissed as complicated. Personally it doesn't seem very complicated to me 🤷‍♂️

ocean90 commented:
We now have #6015 which addresses the remaining concerns. Since this PR was/is only supposed to fix the issue mentioned in #5898 I'm going to merge it so it doesn't miss another release.

Having something translated is more important for translators/testers than trying to figure out how the translations are handled under the hood. The second point can be done in the background (and not hidden in a PR) while visually everything looks fine.

aduth commented:
>The second point can be done in the background (and not hidden in a PR) while visually everything looks fine.

How can we plan to assure this will happen?
